lib/account/account-manager.js

```diff
@@ -83,6 +145,108 @@ module.exports = function (accountService, config, hotp, otDate, promise, random
 
 
     /**
+     * Creates the object for key expiration option. We need to be
+     * specific that we are expiring a file after a given time.
+     *
+     * @param {accountManager~lifeTime} lifeTime
+     * @return {s3~putOptions}
+     * @throws {Error} if we don't pass in something or it's not an object
+     */
+    function keyExpiration(lifeTime) {
+        return otDate.now().plus(lifeTime);
+    }
+
+
+    /**
+     * Finishes logging in a user.
+     *
+     * @param {string} hashedAccountId
+     * @param {string} loginId
+     * @param {accountManager~loginCompleteInfo} loginInfo
+     * @return {Promise.<*>}
+     */
+    function loginCompleteAsync(hashedAccountId, loginId, loginInfo) {
+        return accountService.getLoginFileAsync(hashedAccountId, loginId).then((loginFile) => {
+            if (! secureHash.compare(loginFile.secureHash, loginInfo.password)) {
+                throw new Error("Password hashes do not match");
+            }
+
+            return accountService.getAccountFileAsync(loginFile.accountId).then((accountFile) => {
+                checkMfa(accountFile.mfaKey, loginInfo.currentMfa);
+
+                return accountService.deleteLoginFileAsync(hashedAccountId, loginFile.challengeId);
+            });
+        });
+    }
+
+
+    /**
+     * Maps information we want from the service to return
+     * to the account holder to continue the login process
+     *
+     * @param {accountManager~loginInitiateDetails} details
+     * @return {accountManager~loginInitiateClientInfo}
+     */
+    function loginInitiateDetails(details, salt) {
+        return accountService.getAccountFileAsync(details.accountId).then((accountFile) => {
+            return random.randomIdAsync(config.account.passwordSaltLength).then((passwordSaltId) => {
+                return {
+                    pbkdf2: accountConfig.secureHash,
+                    challenge: {
+                        salt: salt,
```

Comment:
The salt is needed to create the secure hash in the challenge file so we can compare when the client sends in their account hash. This should have been removed when hooking that up, it's left over when we were hashing after they sent in their hash for us to compare.